bundle/config/validate/unique_resource_keys.go

```diff
@@ -0,0 +1,82 @@
+package validate
+
+import (
+	"context"
+	"fmt"
+	"slices"
+
+	"github.com/databricks/cli/bundle"
+	"github.com/databricks/cli/libs/diag"
+	"github.com/databricks/cli/libs/dyn"
+)
+
+// This mutator validates that:
+//
+//  1. Each resource key is unique across different resource types. No two resources
+//     of the same type can have the same key. This is because command like "bundle run"
+//     rely on the resource key to identify the resource to run.
+//     Eg: jobs.foo and pipelines.foo are not allowed simultaneously.
+//
+//  2. Each resource definition is contained within a single file, and is not spread
+//     across multiple files. Note: This is not applicable to resource configuration
+//     defined in a target override. That is why this mutator MUST run before the target
+//     overrides are merged.
+func UniqueResourceKeys() bundle.Mutator {
+	return &uniqueResourceKeys{}
+}
+
+type uniqueResourceKeys struct{}
+
+func (m *uniqueResourceKeys) Name() string {
+	return "validate:unique_resource_keys"
+}
+
+func (m *uniqueResourceKeys) Apply(ctx context.Context, b *bundle.Bundle) diag.Diagnostics {
+	diags := diag.Diagnostics{}
+
+	// Maps of resource key to the paths and locations the resource is defined at.
+	pathsByKey := map[string][]dyn.Path{}
+	locationsByKey := map[string][]dyn.Location{}
+
+	rv := b.Config.Value().Get("resources")
+
+	// return early if no resources are defined or the resources block is empty.
+	if rv.Kind() == dyn.KindInvalid || rv.Kind() == dyn.KindNil {
+		return diags
+	}
+
+	// Gather the paths and locations of all resources.
+	_, err := dyn.MapByPattern(
+		rv,
+		dyn.NewPattern(dyn.AnyKey(), dyn.AnyKey()),
+		func(p dyn.Path, v dyn.Value) (dyn.Value, error) {
+			// The key for the resource. Eg: "my_job" for jobs.my_job.
+			k := p[1].Key()
+
+			// dyn.Path under the hood is a slice. So, we need to clone it.
+			pathsByKey[k] = append(pathsByKey[k], slices.Clone(p))
```

Comment:
you can use dyn.Path.Append, it does work correctly as it copies internally

Reply:
`dyn.Path.Append` is used to combine two paths into a single path. In this case we are tracking two separate paths by tracking them in a `[]dyn.Path`

Reply:
@shreyas-goenka but why do you need clone then? You don't seem to modify the path anywhere anyway

Reply:
The code is not correct in that case. We would end up appending the pointer p(ie slice) here, the value to which it points is changed upstream as we walk the configuration tree.

```
func(p dyn.Path, v dyn.Value) (dyn.Value, error) {
			// The key for the resource. Eg: "my_job" for jobs.my_job.
```

The value `p` here is a pointer, and the underlying value to it is the `prefix` dyn.Path in the `visit` functions, which reuses the same pointer apparently/